File: src/goryu/ofproto/ofmatch.py
# -*- coding: utf-8 -*-

# Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
FIBC Ryu message helper functions
"""

from goryu.ofproto import ofproto

# pylint: disable=too-many-public-methods
class Match(dict):
    """
    Match
    """
    def _put(self, name, value):
        """
        Set match field
        """
        self[name] = value
        return self

    def in_port(self, port):
        """
        in_port=<num>
        """
        return self._put("in_port", port)

    def in_phy_port(self, port):
        """
        in_phy_port=<num>
        """
        return self._put("in_phy_port", port)

    def metadata(self, metadata):
        """
        metadata=<num> or "metadata/mask"
        """
        return self._put("metadata", metadata)

    def eth_dst(self, eth_dst):
        """
        eth_dst="hwaddr" or "hwaddr/mask"
        """
        return self._put("eth_dst", eth_dst)

    def eth_src(self, eth_src):
        """
        eth_src="hwaddr" or "hwaddr/mask"
        """
        return self._put("eth_src", eth_src)

    def eth_type(self, eth_type):
        """
        eth_type=<num>
        """
        return self._put("eth_type", eth_type)

    def vlan_vid(self, vlan_vid, mask=0):
        """
        vlan_vid=<decimal>  (PRESENT BIT added by ofctl)
        vlan_vid="decimal"  (PRESENT BIT added by ofctl)
        vlan_vid="0x..."    (PRESENT BIT not added)
        vlan_vid="vid/mask" (PRESENT BIT not added)
        """
        if not mask:
            if vlan_vid == 0:
                return self._put("vlan_vid", "0x0")

            return self._put("vlan_vid", vlan_vid)

        return self._put("vlan_vid", "{0}/{1}".format(vlan_vid, mask))

    def vlan_pcp(self, vlan_pcp):
        """
        vlan_pcp=<num>
        """
        return self._put("vlan_pcp", vlan_pcp)

    def ip_dscp(self, ip_dscp):
        """
        ip_dscp=<num>
        """
        return self._put("ip_dscp", ip_dscp)

    def ip_ecn(self, ip_ecn):
        """
        ip_ecn=<num>
        """
        return self._put("ip_ecn", ip_ecn)

    def ip_proto(self, ip_proto):
        """
        ip_proto=<num>
        """
        return self._put("ip_proto", ip_proto)

    def ipv4_src(self, ipv4_src):
        """
        ipv4_src="ip" or "ip/<num>" or "ip/mask"
        """
        return self._put("ipv4_src", ipv4_src)

    def ipv4_dst(self, ipv4_dst):
        """
        ipv4_dst="ip" or "ip/<num>" or "ip/mask"
        """
        return self._put("ipv4_dst", ipv4_dst)

    def tcp_src(self, tcp_src):
        """
        tcp_src=<num>
        """
        return self._put("tcp_src", tcp_src)

    def tcp_dst(self, tcp_dst):
        """
        tcp_dst=<num>
        """
        return self._put("tcp_dst", tcp_dst)

    def udp_src(self, udp_src):
        """
        udp_src=<num>
        """
        return self._put("udp_src", udp_src)

    def udp_dst(self, udp_dst):
        """
        udp_dst=<num>
        """
        return self._put("udp_dst", udp_dst)

    def sctp_src(self, sctp_src):
        """
        sctp_src=<num>
        """
        return self._put("sctp_src", sctp_src)

    def sctp_dst(self, sctp_dst):
        """
        sctp_dst=<num>
        """
        return self._put("sctp_dst", sctp_dst)

    def icmpv4_type(self, icmpv4_type):
        """
        icmpv4_type=<num>
        """
        return self._put("icmpv4_type", icmpv4_type)

    def icmpv4_code(self, icmpv4_code):
        """
        icmpv4_code=<num>
        """
        return self._put("icmpv4_code", icmpv4_code)

    def arp_op(self, arp_op):
        """
        arp_op=<num>
        """
        return self._put("arp_op", arp_op)

    def arp_spa(self, arp_spa):
        """
        arp_spa="ip" or "ip/<num>" or "ip/mask"
        """
        self._put("arp_spa", arp_spa)

    def arp_tpa(self, arp_tpa):
        """
        arp_tpa="ip" or "ip/<num>" or "ip/mask"
        """
        self._put("arp_tpa", arp_tpa)

    def arp_sha(self, arp_sha):
        """
        arp_sha="hwaddr" or "hwaddr/mask"
        """
        self._put("arp_sha", arp_sha)

    def arp_tha(self, arp_tha):
        """
        arp_tha="hwaddr" or "hwaddr/mask"
        """
        self._put("arp_tha", arp_tha)

    def ipv6_src(self, ipv6_src):
        """
        ipv6_src="ip" or "ip/<num>" or "ip/mask"
        """
        return self._put("ipv6_src", ipv6_src)

    def ipv6_dst(self, ipv6_dst):
        """
        ipv6_dst="ip" or "ip/<num>" or "ip/mask"
        """
        return self._put("ipv6_dst", ipv6_dst)

    def ipv6_flabel(self, ipv6_flabel):
        """
        ipv6_flabel=<num>
        """
        return self._put("ipv6_flabel", ipv6_flabel)

    def icmpv6_type(self, icmpv6_type):
        """
        icmpv6_type=<num>
        """
        return self._put("icmpv6_type", icmpv6_type)

    def icmpv6_code(self, icmpv6_code):
        """
        icmpv6_code=<num>
        """
        return self._put("icmpv6_code", icmpv6_code)

    def ipv6_nd_target(self, ipv6_nd_target):
        """
        ipv6_nd_target="ip" or "ip/<num>" or "ip/mask"
        """
        return self._put("ipv6_nd_target", ipv6_nd_target)

    def ipv6_nd_sll(self, ipv6_nd_sll):
        """
        ipv6_nd_sll="hwaddr" or "hwaddr/mask"
        """
        return self._put("ipv6_nd_sll", ipv6_nd_sll)

    def ipv6_nd_tll(self, ipv6_nd_tll):
        """
        ipv6_nd_tll="hwaddr" or "hwaddr/mask"
        """
        return self._put("ipv6_nd_tll", ipv6_nd_tll)

    def mpls_label(self, mpls_label):
        """
        mpls_label=<num>
        """
        return self._put("mpls_label", mpls_label)

    def mpls_tc(self, mpls_tc):
        """
        mpls_tc=<num>
        """
        return self._put("mpls_tc", mpls_tc)

    def mpls_bos(self, mpls_bos):
        """
        mpls_bos=<num>
        """
        def _fix_bos():
            return 1 if mpls_bos else 0
        return self._put("mpls_bos", _fix_bos())

    def pbb_isid(self, pbb_isid):
        """
        pbb_isid=<num> or "num" or "num/mask"
        """
        return self._put("pbb_isid", pbb_isid)

    def tunnel_id(self, tunnel_id):
        """
        tunnel_id=<num> or "num" or "num/mask"
        """
        return self._put("tunnel_id", tunnel_id)

    def ipv6_exthdr(self, ipv6_exthdr):
        """
        ipv6_exthdr=<num> or "num" or "num/mask"
        """
        return self._put("ipv6_exthdr", ipv6_exthdr)

    def vrf(self, vrf, use_metadata):
        """
        vrf=<num>
        """
        if use_metadata:
            return self.metadata("{0}/{1}".format(
                vrf << ofproto.VRF_METADATA_SHIFT, ofproto.VRF_METADATA_MASK))

        if vrf:
            return self._put("vrf", vrf)

        return self

    def mpls_type(self, mpls_type, use_metadata):
        """
        mpls_type=<num>
        """
        if use_metadata:
            return self.metadata("{0}/{1}".format(
                mpls_type << ofproto.MPLSTYPE_METADATA_SHIFT, ofproto.MPLSTYPE_METADATA_MASK))

        return self._put("mpls_type", mpls_type)

    def ip_dst(self, ip_dst):
        """
        ip_dst=<ipv4>/<mask> or <ipv6>/<mask>
        """
        if ":" in str(ip_dst):
            return self.eth_type(0x86dd).ipv6_dst(ip_dst)

        return self.eth_type(0x0800).ipv4_dst(ip_dst)

    def ip_src(self, ip_src):
        """
        ip_dst=<ipv4>/<mask> or <ipv6>/<mask>
        """
        if ":" in str(ip_src):
            return self.eth_type(0x86dd).ipv6_src(ip_src)

        return self.eth_type(0x0800).ipv4_src(ip_src)

File: src/goryu/ofproto/ofmatch.go
// -*- coding: utf-8 -*-

// Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ofproto

import (
	"fmt"
	"strings"
)

type Match map[string]interface{}

func (m Match) String() string {
	slist := []string{}

	for name, value := range m {
		switch value.(type) {
		case float64:
			slist = append(slist, fmt.Sprintf("%s=0x%x", name, int64(value.(float64))))
		case string:
			slist = append(slist, fmt.Sprintf("%s=\"%s\"", name, value))
		default:
			slist = append(slist, fmt.Sprintf("%s='%v'", name, value))
		}
	}
	return fmt.Sprintf("[%s]", strings.Join(slist, ","))
}

func (m *Match) Set(name string, value interface{}) *Match {
	(*m)[name] = value
	return m
}

func (m *Match) InPort(port uint32) *Match {
	return m.Set("in_port", port)
}

func (m *Match) InPhyPort(port uint32) *Match {
	return m.Set("in_phy_port", port)
}

func (m *Match) MetadataAndMask(metadata uint64, metadata_mask uint64) *Match {
	return m.Set("metadata", fmt.Sprintf("%d/%d", metadata, metadata_mask))
}

func (m *Match) Metadata(metadata uint64) *Match {
	return m.Set("metadata", metadata)
}

func (m *Match) EthDst(ethDst string) *Match {
	return m.Set("eth_dst", ethDst)
}

func (m *Match) EthSrc(ethSrc string) *Match {
	return m.Set("eth_src", ethSrc)
}

func (m *Match) EthType(ethType uint16) *Match {
	return m.Set("eth_type", ethType)
}

func (m *Match) VlanVidAndMask(vlanVid uint16, mask uint16) *Match {
	return m.Set("vlan_vid", fmt.Sprintf("%d/%d", vlanVid, mask))
}

func (m *Match) VlanVid(vlanVid uint16) *Match {
	if vlanVid == 0 {
		return m.Set("vlan_vid", "0x0")
	}

	return m.Set("vlan_vid", vlanVid)
}

func (m *Match) VlanPCP(pcp uint8) *Match {
	return m.Set("vlan_pcp", pcp)
}

func (m *Match) IPDscp(dscp uint8) *Match {
	return m.Set("ip_dscp", dscp)
}

func (m *Match) IPEcn(ecn uint8) *Match {
	return m.Set("ip_ecn", ecn)
}

func (m *Match) IPProto(proto uint8) *Match {
	return m.Set("ip_proto", proto)
}

func (m *Match) IPv4Src(ipv4Src string) *Match {
	return m.Set("ipv4_src", ipv4Src)
}

func (m *Match) IPv4Dst(ipv4Dst string) *Match {
	return m.Set("ipv4_dst", ipv4Dst)
}

func (m *Match) TCPSrc(tcpSrc uint16) *Match {
	return m.Set("tcp_src", tcpSrc)
}

func (m *Match) TCPDst(tcpDst uint16) *Match {
	return m.Set("tcp_dst", tcpDst)
}

func (m *Match) UDPSrc(udpSrc uint16) *Match {
	return m.Set("udp_src", udpSrc)
}

func (m *Match) UDPDst(udpDst uint16) *Match {
	return m.Set("udp_dst", udpDst)
}

func (m *Match) SctpSrc(sctpSrc uint16) *Match {
	return m.Set("sctp_src", sctpSrc)
}

func (m *Match) SctpDst(sctpDst uint16) *Match {
	return m.Set("sctp_dst", sctpDst)
}

func (m *Match) ICMPv4Code(code uint8) *Match {
	return m.Set("icmpv4_code", code)
}

func (m *Match) ICMPv4Type(t uint8) *Match {
	return m.Set("icmpv4_type", t)
}

func (m *Match) ARPOp(op uint16) *Match {
	return m.Set("arp_op", op)
}

func (m *Match) ARPSpa(spa string) *Match {
	return m.Set("arp_spa", spa)
}

func (m *Match) ARPTpa(tpa string) *Match {
	return m.Set("arp_tpa", tpa)
}

func (m *Match) ARPSha(sha string) *Match {
	return m.Set("arp_sha", sha)
}

func (m *Match) ARPTha(tha string) *Match {
	return m.Set("arp_tha", tha)
}

func (m *Match) IPv6Src(ipv6Src string) *Match {
	return m.Set("ipv6_src", ipv6Src)
}

func (m *Match) IPv6Dst(ipv6Dst string) *Match {
	return m.Set("ipv6_dst", ipv6Dst)
}

func (m *Match) IPv6Flabel(flabel uint32) *Match {
	return m.Set("ipv6_flabel", flabel)
}

func (m *Match) ICMPv6Code(code uint8) *Match {
	return m.Set("icmpv6_code", code)
}

func (m *Match) ICMPv6Type(t uint8) *Match {
	return m.Set("icmpv6_type", t)
}

func (m *Match) IPv6NDTarger(ipv6NDTarget string) *Match {
	return m.Set("ipv6_nd_target", ipv6NDTarget)
}

func (m *Match) IPv6NDSll(ipv6NDSll string) *Match {
	return m.Set("ipv6_nd_sll", ipv6NDSll)
}

func (m *Match) IPv6NDTll(ipv6NDTll string) *Match {
	return m.Set("ipv6_nd_tll", ipv6NDTll)
}

func (m *Match) MPLSLabel(label uint32) *Match {
	return m.Set("mpls_label", label)
}

func (m *Match) MPLSTc(tc uint8) *Match {
	return m.Set("mpls_tc", tc)
}

func (m *Match) MPLSBos(bos uint8) *Match {
	return m.Set("mpls_bos", bos)
}

func (m *Match) PBBIsid(isid uint32) *Match {
	return m.Set("pbb_isid", isid)
}

func (m *Match) PBBIsidAndMask(isid, mask uint32) *Match {
	return m.Set("pbb_isid", fmt.Sprintf("%d/%d", isid, mask))
}

func (m *Match) TunnelId(tunnelId uint64) *Match {
	return m.Set("tunnel_id", tunnelId)
}

func (m *Match) TunnelIdAndMask(tunnelId, mask uint64) *Match {
	return m.Set("tunnel_id", fmt.Sprintf("%d/%d", tunnelId, mask))
}

func (m *Match) IPv6ExtHdr(exthdr uint16) *Match {
	return m.Set("ipv6_exthdr", exthdr)
}

func (m *Match) IPv6ExtHdrAndMask(exthdr, mask uint16) *Match {
	return m.Set("ipv6_exthdr", fmt.Sprintf("%d/%d", exthdr, mask))
}

func (m *Match) Vrf(vrf uint8, useMetadata bool) *Match {
	if useMetadata {
		return m.MetadataAndMask(uint64(vrf)<<VRF_METADATA_SHIFT, VRF_METADATA_MASK)
	} else {
		return m.Set("vrf", vrf)
	}
}

func (m *Match) MPLSType(mplsType uint8, useMetadata bool) *Match {
	if useMetadata {
		return m.MetadataAndMask(uint64(mplsType)<<MPLSTYPE_METADATA_SHIFT, MPLSTYPE_METADATA_MASK)
	} else {
		return m.Set("mpls_type", mplsType)
	}
}

func (m *Match) IPDst(ipDst string) *Match {
	if strings.Contains(ipDst, ":") {
		return m.EthType(0x86dd).IPv6Dst(ipDst)
	} else {
		return m.EthType(0x0800).IPv4Dst(ipDst)
	}
}

func (m *Match) IPSrc(ipSrc string) *Match {
	if strings.Contains(ipSrc, ":") {
		return m.EthType(0x86dd).IPv6Src(ipSrc)
	} else {
		return m.EthType(0x0800).IPv4Src(ipSrc)
	}
}
